File: src/bridge.ts
import { ChildProcessWithoutNullStreams, spawn } from 'node:child_process';
import * as readline from 'node:readline';
import * as vscode from 'vscode';

/** Chat request forwarded by the Rust bridge, mirroring `protocol::ChatRequest`. */
interface ChatRequest {
	model?: string;
	messages: ChatMessage[];
	tools?: ToolDef[];
	temperature?: number;
	jsonOutput?: boolean;
}

interface ChatMessage {
	role: 'system' | 'user' | 'assistant' | 'tool';
	content: string;
	toolCalls?: { id: string; name: string; arguments: string }[];
	toolCallId?: string;
}

interface ToolDef {
	name: string;
	description: string;
	inputSchema: object;
}

type HostBound =
	| { type: 'ready'; port: number; token: string }
	| ({ type: 'request'; id: string } & ChatRequest)
	| { type: 'cancel'; id: string }
	| { type: 'tool'; id: string; name: string; arguments: Record<string, unknown> }
	| { type: 'log'; level: string; message: string };

export interface BridgeEndpoint {
	port: number;
	token: string;
	baseUrl: string;
	/** Streamable HTTP endpoint VS Code connects the MCP client to. */
	mcpUrl: string;
}

/** Runs one MCP tool invocation on behalf of the bridge. */
export type ToolHandler = (name: string, args: Record<string, unknown>) => Promise<string>;

const JSON_ONLY_INSTRUCTION =
	'Reply with a single valid JSON document and nothing else: no prose, no explanation, no markdown code fences.';

/** How many times a turn that streamed nothing is replayed before it is reported. */
const EMPTY_RESPONSE_ATTEMPTS = 3;
/** Multiplied by the attempt number, so the provider gets a widening pause. */
const EMPTY_RESPONSE_BACKOFF_MS = 2000;
/** Below this, the turn was refused locally rather than answered with nothing. */
const EMPTY_RESPONSE_MIN_RETRY_MS = 1000;
/** How long to stop calling the model once Copilot starts refusing this extension. */
const BLOCKED_COOLDOWN_MS = 120_000;
const BLOCKED_MESSAGE =
	'GitHub Copilot has temporarily blocked this extension for sending too many requests. ' +
	'Every review turn will fail until the block clears, so wait a few minutes before ' +
	'retrying, and raise sashiko.requestIntervalMs to spread requests further apart.';

/**
 * Owns the `sashiko-vscode-bridge serve` child process and answers its completion
 * requests with `vscode.lm`. This is the only place the extension touches the
 * language model; everything else in the pipeline is Rust.
 */
export class LanguageModelBridge implements vscode.Disposable {
	private child?: ChildProcessWithoutNullStreams;
	private endpoint?: BridgeEndpoint;
	private ready?: Promise<BridgeEndpoint>;
	private toolHandler?: ToolHandler;
	private blockedUntil = 0;
	private nextRequestAt = 0;
	private readonly inflight = new Map<string, vscode.CancellationTokenSource>();
	private readonly endpointChanged = new vscode.EventEmitter<void>();

	/** Fires whenever the listener moves, so the MCP server definition can be refreshed. */
	readonly onDidChangeEndpoint = this.endpointChanged.event;

	constructor(
		private readonly binaryPath: string,
		private readonly log: vscode.LogOutputChannel
	) {}

	setToolHandler(handler: ToolHandler): void {
		this.toolHandler = handler;
	}

	/** Starts the bridge if needed and resolves once the HTTP endpoint is listening. */
	async start(): Promise<BridgeEndpoint> {
		if (this.endpoint) {
			return this.endpoint;
		}
		this.ready ??= this.spawnBridge();
		return this.ready;
	}

	private spawnBridge(): Promise<BridgeEndpoint> {
		return new Promise<BridgeEndpoint>((resolve, reject) => {
			let child: ChildProcessWithoutNullStreams;
			try {
				child = spawn(this.binaryPath, ['serve'], { stdio: 'pipe' });
			} catch (error) {
				reject(error);
				return;
			}
			this.child = child;

			child.on('error', (error) => {
				this.log.error(`bridge failed to start: ${error.message}`);
				this.reset();
				reject(error);
			});
			child.on('exit', (code, signal) => {
				this.log.warn(`bridge exited (code=${code ?? 'null'}, signal=${signal ?? 'null'})`);
				this.reset();
			});

			readline.createInterface({ input: child.stderr }).on('line', (line) => {
				this.log.warn(`bridge: ${line}`);
			});

			readline.createInterface({ input: child.stdout }).on('line', (line) => {
				let message: HostBound;
				try {
					message = JSON.parse(line) as HostBound;
				} catch {
					this.log.warn(`bridge emitted non-JSON output: ${line}`);
					return;
				}
				this.handle(message, resolve);
			});
		});
	}

	private handle(message: HostBound, onReady: (endpoint: BridgeEndpoint) => void): void {
		switch (message.type) {
			case 'ready': {
				const endpoint: BridgeEndpoint = {
					port: message.port,
					token: message.token,
					baseUrl: `http://127.0.0.1:${message.port}/v1`,
					mcpUrl: `http://127.0.0.1:${message.port}/mcp`
				};
				this.endpoint = endpoint;
				this.log.info(`bridge listening on ${endpoint.baseUrl}`);
				void this.publishModels();
				onReady(endpoint);
				this.endpointChanged.fire();
				break;
			}
			case 'log':
				this.log.appendLine(`[bridge] ${message.level}: ${message.message}`);
				break;
			case 'cancel':
				this.inflight.get(message.id)?.cancel();
				break;
			case 'tool':
				void this.runTool(message.id, message.name, message.arguments);
				break;
			case 'request':
				void this.answer(message.id, message);
				break;
		}
	}

	private async runTool(
		id: string,
		name: string,
		args: Record<string, unknown>
	): Promise<void> {
		if (!this.toolHandler) {
			this.send({ type: 'toolResult', id, text: 'Sashiko tools are not ready yet.', isError: true });
			return;
		}
		try {
			this.send({ type: 'toolResult', id, text: await this.toolHandler(name, args), isError: false });
		} catch (error) {
			const text = error instanceof Error ? error.message : String(error);
			this.log.error(`tool ${name} failed: ${text}`);
			this.send({ type: 'toolResult', id, text, isError: true });
		}
	}

	private send(payload: unknown): void {
		this.child?.stdin.write(`${JSON.stringify(payload)}\n`);
	}

	/** Publishes the model catalogue so Sashiko's `GET /v1/models` returns something useful. */
	async publishModels(): Promise<void> {
		const models = await vscode.lm.selectChatModels();
		this.send({
			type: 'models',
			models: models.map((model) => ({
				id: model.id,
				name: model.name,
				vendor: model.vendor,
				family: model.family,
				maxInputTokens: model.maxInputTokens
			}))
		});
	}

	private async answer(id: string, request: ChatRequest): Promise<void> {
		const cancellation = new vscode.CancellationTokenSource();
		this.inflight.set(id, cancellation);
		try {
			if (Date.now() < this.blockedUntil) {
				this.send({ type: 'error', id, message: BLOCKED_MESSAGE });
				return;
			}
			const model = await resolveModel(request.model);
			const options: vscode.LanguageModelChatRequestOptions = {
				justification: 'Sashiko is reviewing Linux kernel patches on your behalf.'
			};
			if (request.tools?.length) {
				options.tools = request.tools.map((tool) => ({
					name: tool.name,
					description: tool.description,
					inputSchema: tool.inputSchema
				}));
				options.toolMode = vscode.LanguageModelChatToolMode.Auto;
			}

			const messages = toChatMessages(request.messages, request.jsonOutput);
			// Sashiko builds its own prompts and does not always honour the budget in
			// Settings.toml, so an oversized turn would otherwise fail without a reason.
			const tokens = await countPromptTokens(model, messages, cancellation.token);
			if (tokens > model.maxInputTokens) {
				const message =
					`The prompt is ${tokens.toLocaleString()} tokens but '${model.id}' accepts at most ` +
					`${model.maxInputTokens.toLocaleString()}. Pick a model with a larger context window, ` +
					'or lower sashiko.maxInputTokens so Sashiko sends less context.';
				this.log.error(message);
				this.send({ type: 'error', id, message });
				return;
			}

			await this.waitForSlot(cancellation.token);

			// A model that streams nothing is usually a transient provider hiccup. Sashiko
			// retries with no backoff at all, so absorb it here rather than let a whole
			// review collapse into a burst of failed turns.
			const shape =
				`${messages.length} messages, ${tokens.toLocaleString()}/` +
				`${model.maxInputTokens.toLocaleString()} tokens, ${options.tools?.length ?? 0} tools`;
			for (let attempt = 1; ; attempt++) {
				const started = Date.now();
				const response = await model.sendRequest(messages, options, cancellation.token);

				let toolCalls = 0;
				let textLength = 0;
				const ignored = new Set<string>();
				for await (const part of response.stream) {
					if (part instanceof vscode.LanguageModelTextPart) {
						textLength += part.value.length;
						this.send({ type: 'chunk', id, delta: part.value });
					} else if (part instanceof vscode.LanguageModelToolCallPart) {
						toolCalls += 1;
						this.send({
							type: 'toolCall',
							id,
							call: {
								id: part.callId,
								name: part.name,
								arguments: JSON.stringify(part.input ?? {})
							}
						});
					} else {
						// Reasoning models stream parts this bridge has no wire format for.
						ignored.add((part as object)?.constructor?.name ?? 'unknown');
					}
				}

				if (textLength > 0 || toolCalls > 0) {
					if (ignored.size > 0) {
						this.log.debug(`ignored response parts: ${[...ignored].join(', ')}`);
					}
					this.send({ type: 'done', id, finishReason: toolCalls > 0 ? 'tool_calls' : 'stop' });
					return;
				}

				// Nothing was streamed, so nothing was forwarded either and the turn can be
				// replayed safely. Sashiko parses the body as JSON, so giving up here at
				// least surfaces the real cause instead of a parse error several retries on.
				const elapsed = Date.now() - started;
				const detail = ignored.size > 0 ? ` (only ${[...ignored].join(', ')} parts)` : '';
				const message =
					`The language model '${model.id}' returned an empty response${detail} ` +
					`after ${elapsed}ms.`;
				// An instant empty stream is the provider refusing the turn without asking the
				// model, so replaying it only spends quota against whatever refused it.
				const worthRetrying = elapsed >= EMPTY_RESPONSE_MIN_RETRY_MS;
				if (!worthRetrying && (await this.isRefusingTurns(model))) {
					this.blockedUntil = Date.now() + BLOCKED_COOLDOWN_MS;
					this.log.error(BLOCKED_MESSAGE);
					this.send({ type: 'error', id, message: BLOCKED_MESSAGE });
					return;
				}
				if (
					!worthRetrying ||
					attempt >= EMPTY_RESPONSE_ATTEMPTS ||
					cancellation.token.isCancellationRequested
				) {
					this.log.warn(`${message} [${shape}]`);
					this.send({ type: 'error', id, message });
					return;
				}
				this.log.warn(`${message} [${shape}] Retrying (${attempt}/${EMPTY_RESPONSE_ATTEMPTS - 1}).`);
				await delay(attempt * EMPTY_RESPONSE_BACKOFF_MS, cancellation.token);
			}
		} catch (error) {
			const message = describeError(error);
			this.log.error(`language model request failed: ${message}`);
			this.send({ type: 'error', id, message });
		} finally {
			this.inflight.delete(id);
			cancellation.dispose();
		}
	}

	/** Spaces requests out: Copilot temporarily blocks extensions that burst. */
	private async waitForSlot(token: vscode.CancellationToken): Promise<void> {
		const interval = vscode.workspace
			.getConfiguration('sashiko')
			.get<number>('requestIntervalMs', 250);
		if (interval <= 0) {
			return;
		}
		const now = Date.now();
		const slot = Math.max(now, this.nextRequestAt);
		this.nextRequestAt = slot + interval;
		if (slot > now) {
			await delay(slot - now, token);
		}
	}

	/**
	 * Sends one trivial turn to tell a rejected prompt apart from a refused extension.
	 * Copilot blocks an extension that sends too many requests and VS Code surfaces that
	 * as an empty stream rather than an error, so a tiny prompt failing too is the tell.
	 */
	private async isRefusingTurns(model: vscode.LanguageModelChat): Promise<boolean> {
		const cancellation = new vscode.CancellationTokenSource();
		const started = Date.now();
		try {
			const response = await model.sendRequest(
				[vscode.LanguageModelChatMessage.User('Reply with OK.')],
				{ justification: 'Sashiko is checking whether the language model still answers.' },
				cancellation.token
			);
			let text = '';
			for await (const part of response.stream) {
				if (part instanceof vscode.LanguageModelTextPart) {
					text += part.value;
				}
			}
			this.log.warn(`probe: '${model.id}' returned ${text.length} chars in ${Date.now() - started}ms`);
			return text.length === 0;
		} catch (error) {
			this.log.warn(`probe: '${model.id}' failed after ${Date.now() - started}ms: ${describeError(error)}`);
			return false;
		} finally {
			cancellation.dispose();
		}
	}

	private reset(): void {
		for (const cancellation of this.inflight.values()) {
			cancellation.cancel();
			cancellation.dispose();
		}
		this.inflight.clear();
		this.child = undefined;
		this.endpoint = undefined;
		this.ready = undefined;
	}

	dispose(): void {
		this.child?.kill();
		this.reset();
		this.endpointChanged.dispose();
	}
}

/**
 * Resolves the model Sashiko asked for, falling back to any available chat model so a
 * stale id in Settings.toml does not break the review.
 */
export async function resolveModel(id?: string): Promise<vscode.LanguageModelChat> {
	if (id) {
		const exact = await vscode.lm.selectChatModels({ id });
		if (exact.length > 0) {
			return exact[0];
		}
		const byFamily = await vscode.lm.selectChatModels({ family: id });
		if (byFamily.length > 0) {
			return byFamily[0];
		}
	}
	const any = await vscode.lm.selectChatModels();
	if (any.length === 0) {
		throw new Error(
			'No language model is available in VS Code. Sign in to GitHub Copilot (or another chat provider) and try again.'
		);
	}
	return any[0];
}

/**
 * Maps OpenAI-shaped messages onto the VS Code chat API. The API has no system role,
 * so system prompts are folded into the leading user turn.
 */
function toChatMessages(
	messages: ChatMessage[],
	jsonOutput?: boolean
): vscode.LanguageModelChatMessage[] {
	const result: vscode.LanguageModelChatMessage[] = [];

	for (const message of messages) {
		switch (message.role) {
			case 'assistant': {
				const parts: (vscode.LanguageModelTextPart | vscode.LanguageModelToolCallPart)[] = [];
				if (message.content) {
					parts.push(new vscode.LanguageModelTextPart(message.content));
				}
				for (const call of message.toolCalls ?? []) {
					parts.push(
						new vscode.LanguageModelToolCallPart(call.id, call.name, parseArguments(call.arguments))
					);
				}
				if (parts.length > 0) {
					result.push(vscode.LanguageModelChatMessage.Assistant(parts));
				}
				break;
			}
			case 'tool': {
				if (!message.toolCallId) {
					break;
				}
				result.push(
					vscode.LanguageModelChatMessage.User([
						new vscode.LanguageModelToolResultPart(message.toolCallId, [
							new vscode.LanguageModelTextPart(message.content)
						])
					])
				);
				break;
			}
			default: {
				if (message.content.trim()) {
					result.push(vscode.LanguageModelChatMessage.User(message.content));
				}
				break;
			}
		}
	}

	if (result.length === 0) {
		result.push(vscode.LanguageModelChatMessage.User('Continue.'));
	}
	// The chat API has no equivalent of OpenAI's `response_format`, so the only way to
	// hold the model to JSON is to say so in the prompt.
	if (jsonOutput) {
		result.push(vscode.LanguageModelChatMessage.User(JSON_ONLY_INSTRUCTION));
	}
	return result;
}

/** Sleeps between retries, waking early when the request is cancelled. */
function delay(ms: number, token: vscode.CancellationToken): Promise<void> {
	return new Promise<void>((resolve) => {
		const done = (): void => {
			clearTimeout(timer);
			subscription.dispose();
			resolve();
		};
		const timer = setTimeout(done, ms);
		const subscription = token.onCancellationRequested(done);
	});
}

/** Adds up the prompt's token count, which the chat API only reports per message. */
async function countPromptTokens(
	model: vscode.LanguageModelChat,
	messages: vscode.LanguageModelChatMessage[],
	token: vscode.CancellationToken
): Promise<number> {
	const counts = await Promise.all(
		messages.map((message) => model.countTokens(message, token))
	);
	return counts.reduce((total, count) => total + count, 0);
}

function parseArguments(raw: string): object {
	try {
		const parsed: unknown = JSON.parse(raw || '{}');
		return typeof parsed === 'object' && parsed !== null ? parsed : {};
	} catch {
		return {};
	}
}

function describeError(error: unknown): string {
	if (error instanceof vscode.LanguageModelError) {
		return `${error.code}: ${error.message}`;
	}
	return error instanceof Error ? error.message : String(error);
}

File: rust/src/bridge.rs
//! Duplex link to the VS Code extension host over the process' stdio.

use std::collections::HashMap;
use std::sync::{Arc, Mutex};

use anyhow::{anyhow, Result};
use serde_json::Value;
use tokio::io::{AsyncBufReadExt, AsyncWriteExt, BufReader};
use tokio::sync::{mpsc, oneshot};

use crate::protocol::{ChatRequest, ExtensionBound, HostBound, ModelInfo, ToolCall, Usage};

/// A single piece of a streamed completion.
#[derive(Debug, Clone)]
pub enum StreamEvent {
    Text(String),
    ToolCall(ToolCall),
    Done {
        finish_reason: Option<String>,
        usage: Option<Usage>,
    },
    Error(String),
}

/// What the extension host made of an MCP tool invocation.
#[derive(Debug, Clone)]
pub struct ToolOutcome {
    pub text: String,
    pub is_error: bool,
}

#[derive(Default)]
struct State {
    pending: HashMap<String, mpsc::UnboundedSender<StreamEvent>>,
    tools: HashMap<String, oneshot::Sender<ToolOutcome>>,
    models: Vec<ModelInfo>,
}

pub struct Host {
    outbox: mpsc::UnboundedSender<String>,
    state: Mutex<State>,
}

impl Host {
    /// Spawns the stdout writer and stdin reader tasks and returns the link.
    pub fn spawn() -> Arc<Self> {
        let (outbox, mut rx) = mpsc::unbounded_channel::<String>();
        let host = Arc::new(Self {
            outbox,
            state: Mutex::new(State::default()),
        });

        tokio::spawn(async move {
            let mut stdout = tokio::io::stdout();
            while let Some(line) = rx.recv().await {
                if stdout.write_all(line.as_bytes()).await.is_err()
                    || stdout.write_all(b"\n").await.is_err()
                    || stdout.flush().await.is_err()
                {
                    break;
                }
            }
        });

        let reader_host = Arc::clone(&host);
        tokio::spawn(async move {
            let mut lines = BufReader::new(tokio::io::stdin()).lines();
            while let Ok(Some(line)) = lines.next_line().await {
                let line = line.trim();
                if line.is_empty() {
                    continue;
                }
                match serde_json::from_str::<ExtensionBound>(line) {
                    Ok(message) => reader_host.dispatch(message),
                    Err(err) => reader_host.log("warn", format!("unparsable host message: {err}")),
                }
            }
            // stdin closed: the extension host is gone, so is our reason to live.
            reader_host.fail_all("extension host disconnected");
            std::process::exit(0);
        });

        host
    }

    fn dispatch(&self, message: ExtensionBound) {
        match message {
            ExtensionBound::Models { models } => {
                self.state.lock().unwrap().models = models;
            }
            ExtensionBound::Chunk { id, delta } => self.emit(&id, StreamEvent::Text(delta), false),
            ExtensionBound::ToolCall { id, call } => {
                self.emit(&id, StreamEvent::ToolCall(call), false)
            }
            ExtensionBound::Done {
                id,
                finish_reason,
                usage,
            } => self.emit(
                &id,
                StreamEvent::Done {
                    finish_reason,
                    usage,
                },
                true,
            ),
            ExtensionBound::Error { id, message } => {
                self.emit(&id, StreamEvent::Error(message), true)
            }
            ExtensionBound::ToolResult { id, text, is_error } => {
                let sender = self.state.lock().unwrap().tools.remove(&id);
                if let Some(sender) = sender {
                    let _ = sender.send(ToolOutcome { text, is_error });
                }
            }
        }
    }

    fn emit(&self, id: &str, event: StreamEvent, terminal: bool) {
        let mut state = self.state.lock().unwrap();
        let sender = if terminal {
            state.pending.remove(id)
        } else {
            state.pending.get(id).cloned()
        };
        drop(state);
        if let Some(sender) = sender {
            let _ = sender.send(event);
        }
    }

    fn fail_all(&self, reason: &str) {
        let mut state = self.state.lock().unwrap();
        let pending = std::mem::take(&mut state.pending);
        // Dropping the oneshot senders makes every awaiting tool call resolve as an error.
        state.tools.clear();
        drop(state);
        for (_, sender) in pending {
            let _ = sender.send(StreamEvent::Error(reason.to_string()));
        }
    }

    fn send(&self, message: HostBound) {
        match serde_json::to_string(&message) {
            Ok(line) => {
                let _ = self.outbox.send(line);
            }
            Err(err) => eprintln!("failed to encode host message: {err}"),
        }
    }

    pub fn log(&self, level: &'static str, message: impl Into<String>) {
        self.send(HostBound::Log {
            level,
            message: message.into(),
        });
    }

    pub fn announce_ready(&self, port: u16, token: &str) {
        self.send(HostBound::Ready {
            port,
            token: token.to_string(),
        });
    }

    pub fn models(&self) -> Vec<ModelInfo> {
        self.state.lock().unwrap().models.clone()
    }

    /// Forwards a completion request to the extension host and returns its event stream.
    pub fn request(self: &Arc<Self>, payload: ChatRequest) -> ResponseStream {
        let id = uuid::Uuid::new_v4().to_string();
        let (tx, rx) = mpsc::unbounded_channel();
        self.state.lock().unwrap().pending.insert(id.clone(), tx);
        self.send(HostBound::Request {
            id: id.clone(),
            payload,
        });
        ResponseStream {
            host: Arc::clone(self),
            id,
            rx,
            finished: false,
        }
    }

    /// Hands an MCP tool invocation to the extension host and waits for its verdict.
    ///
    /// The tools operate on workspace configuration and the Problems panel, so the
    /// extension host is the only process that can run them.
    pub async fn call_tool(&self, name: &str, arguments: Value) -> Result<ToolOutcome> {
        let id = uuid::Uuid::new_v4().to_string();
        let (tx, rx) = oneshot::channel();
        self.state.lock().unwrap().tools.insert(id.clone(), tx);
        self.send(HostBound::Tool {
            id,
            name: name.to_string(),
            arguments,
        });
        rx.await
            .map_err(|_| anyhow!("the VS Code extension host did not answer the tool call"))
    }
}

/// Stream of events for one in-flight completion.
///
/// Dropping it before completion cancels the request in the extension host, which
/// is what makes HTTP client disconnects propagate all the way to `vscode.lm`.
pub struct ResponseStream {
    host: Arc<Host>,
    id: String,
    rx: mpsc::UnboundedReceiver<StreamEvent>,
    finished: bool,
}

impl ResponseStream {
    pub async fn recv(&mut self) -> Option<StreamEvent> {
        let event = self.rx.recv().await;
        if matches!(
            event,
            None | Some(StreamEvent::Done { .. }) | Some(StreamEvent::Error(_))
        ) {
            self.finished = true;
        }
        event
    }

    /// Drains the stream into a single response body.
    pub async fn collect(mut self) -> Result<(String, Vec<ToolCall>, Option<String>, Usage)> {
        let mut text = String::new();
        let mut calls = Vec::new();
        loop {
            match self.recv().await {
                Some(StreamEvent::Text(delta)) => text.push_str(&delta),
                Some(StreamEvent::ToolCall(call)) => calls.push(call),
                Some(StreamEvent::Done {
                    finish_reason,
                    usage,
                }) => return Ok((text, calls, finish_reason, usage.unwrap_or_default())),
                Some(StreamEvent::Error(message)) => return Err(anyhow!(message)),
                None => return Err(anyhow!("language model stream closed unexpectedly")),
            }
        }
    }
}

impl Drop for ResponseStream {
    fn drop(&mut self) {
        if self.finished {
            return;
        }
        self.host.state.lock().unwrap().pending.remove(&self.id);
        self.host.send(HostBound::Cancel {
            id: self.id.clone(),
        });
    }
}
